File: app/src/main/java/znyoo/name/baseproject/SampleApplicationLike.java
package znyoo.name.baseproject;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.multidex.MultiDex;

import com.blankj.utilcode.util.LogUtils;
import com.coder.zzq.smartshow.core.SmartShow;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.DefaultRefreshFooterCreator;
import com.scwang.smartrefresh.layout.api.DefaultRefreshHeaderCreator;
import com.scwang.smartrefresh.layout.api.RefreshFooter;
import com.scwang.smartrefresh.layout.api.RefreshHeader;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.constant.SpinnerStyle;
import com.scwang.smartrefresh.layout.footer.ClassicsFooter;
import com.scwang.smartrefresh.layout.header.ClassicsHeader;
import com.tencent.bugly.Bugly;
import com.tencent.bugly.beta.Beta;
import com.tencent.bugly.beta.interfaces.BetaPatchListener;
import com.tencent.tinker.entry.DefaultApplicationLike;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.HasAndroidInjector;
import me.jessyan.autosize.AutoSizeConfig;
import znyoo.name.base.base.BaseActivity;
import znyoo.name.base.common.CommonSetKt;
import znyoo.name.baseproject.ui.di.AppInjector;
import znyoo.name.netlibrary.retrofit.OnlineConfig;
import znyoo.name.netlibrary.retrofit.XH_RXOnline;

import static znyoo.name.base.common.CommonSetKt.BUGLY_APPID;
import static znyoo.name.base.common.CommonSetKt.ISTEST;
import static znyoo.name.base.common.UrlConstantKt.MAIN_URL;
import static znyoo.name.base.common.UrlConstantKt.TEST_MAIN_URL;


/**
 * 自定义ApplicationLike类.
 *
 * 注意：这个类是Application的代理类，以前所有在Application的实现必须要全部拷贝到这里<br/>
 *
 *  make your Application implement HasAndroidInjector and @Inject a DispatchingAndroidInjector<Object> to return from the androidInjector() method:
 * @author wenjiewu
 * @since 2016/11/7
 */
public class SampleApplicationLike extends DefaultApplicationLike implements HasAndroidInjector {

    @Inject
    DispatchingAndroidInjector<Object> dispatchingAndroidInjector;

    public static final String TAG = "Tinker.SampleApplicationLike";

    private static SampleApplicationLike instance;

    private List<Activity> activitys;

    //static 代码段可以防止内存泄露
    static {

        //设置全局的Header构建器  recycleview下拉上拉样式
        SmartRefreshLayout.setDefaultRefreshHeaderCreator(new DefaultRefreshHeaderCreator() {
            @NonNull
            @Override
            public RefreshHeader createRefreshHeader(@NonNull Context context, @NonNull RefreshLayout layout) {
                //                layout.setPrimaryColorsId(R.color.blackColor, android.R.color.white);//全局设置主题颜色
                return new ClassicsHeader(context).setSpinnerStyle(SpinnerStyle.Scale);//指定为经典Header，默认是 贝塞尔雷达Header
            }
        });
        SmartRefreshLayout.setDefaultRefreshFooterCreator(new DefaultRefreshFooterCreator() {
            @NonNull
            @Override
            public RefreshFooter createRefreshFooter(@NonNull Context context, @NonNull RefreshLayout layout) {
                //指定为经典Footer，默认是 BallPulseFooter
                return new ClassicsFooter(context).setSpinnerStyle(SpinnerStyle.Translate);
            }
        });
    }

    public SampleApplicationLike(Application application, int tinkerFlags,
                                 boolean tinkerLoadVerifyFlag, long applicationStartElapsedTime,
                                 long applicationStartMillisTime, Intent tinkerResultIntent) {
        super(application, tinkerFlags, tinkerLoadVerifyFlag, applicationStartElapsedTime,
                applicationStartMillisTime, tinkerResultIntent);
        instance = this;
        activitys = new ArrayList<>();
    }


    @Override
    public void onCreate() {
        super.onCreate();
        initialize();

//        AppInjector.INSTANCE.inject(getApplication());
    }


    @TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
    @Override
    public void onBaseContextAttached(Context base) {
        super.onBaseContextAttached(base);
        // you must install multiDex whatever tinker is installed!
        MultiDex.install(base);

        // TODO: 安装tinker
        Beta.installTinker(this);
    }

    @Override
    public void onTerminate() {
        super.onTerminate();
        Beta.unInit();
    }

    /**
     * 单例，返回一个实例
     */
    public static SampleApplicationLike getInstance() {
        return instance;
    }


    // 遍历所有Activity并finish
    public void addActivity(Activity activity) {
        if (activitys == null) {
            return;
        }
        if (activitys.size() > 0) {
            if (!activitys.contains(activity)) {
                activitys.add(activity);
            }
        } else {
            activitys.add(activity);
        }
    }

    // 遍历所有Activity并finish
    public void exit() {
        if (activitys != null && activitys.size() > 0) {
            for (Activity activity : activitys) {
                activity.finish();
            }
        }
        System.exit(0);
    }

    private void initialize(){
        Beta.initDelay = 0;
        Beta.upgradeCheckPeriod = 0;
        Beta.canShowUpgradeActs.add(BaseActivity.class);
        // 设置是否开启热更新能力，默认为true
        Beta.enableHotfix = true;
        // 设置是否自动下载补丁，默认为true
        Beta.canAutoDownloadPatch = true;
        // 设置是否自动合成补丁，默认为true
        Beta.canAutoPatch = true;
        // 设置是否提示用户重启，默认为false
        Beta.canNotifyUserRestart = true;
        // 补丁回调接口
        Beta.betaPatchListener = new BetaPatchListener() {
            @Override
            public void onPatchReceived(String patchFile) {
                LogUtils.e("补丁下载地址" + patchFile);
            }

            @Override
            public void onDownloadReceived(long savedLength, long totalLength) {
                LogUtils.e(String.format(Locale.getDefault(), "%s %d%%",
                        Beta.strNotificationDownloading,
                        (int) (totalLength == 0 ? 0 : savedLength * 100 / totalLength)));
            }

            @Override
            public void onDownloadSuccess(String msg) {
                LogUtils.e("补丁下载成功");
            }

            @Override
            public void onDownloadFailure(String msg) {
                LogUtils.e("补丁下载失败");
            }

            @Override
            public void onApplySuccess(String msg) {
                LogUtils.e("补丁应用成功");
            }

            @Override
            public void onApplyFailure(String msg) {
                LogUtils.e("补丁应用失败");
            }

            @Override
            public void onPatchRollback() {

            }
        };
        // 设置开发设备，默认为false，上传补丁如果下发范围指定为“开发设备”，需要调用此接口来标识开发设备
        Bugly.setIsDevelopmentDevice(getApplication(), false);
        // 多渠道需求塞入
        // String channel = WalleChannelReader.getChannel(getApplication());
        // Bugly.setAppChannel(getApplication(), channel);
        // 这里实现SDK初始化，appId替换成你的在Bugly平台申请的appId
        Bugly.init(getApplication(), BUGLY_APPID, false);
        Beta.checkUpgrade(false,false);


        //rutosize
//        AutoSizeConfig.getInstance().setCustomFragment(true);
        SmartShow.init(getApplication());

        initOnline();
    }

    public void initOnline() {
        OnlineConfig config = new OnlineConfig();
        config.url = CommonSetKt.ISTEST? TEST_MAIN_URL : MAIN_URL;
        config.isShowWait = false;
        XH_RXOnline.getInstance().setOnlineConfig(config);
    }

    @Override
    public AndroidInjector<Object> androidInjector() {
        return dispatchingAndroidInjector;
    }
}

File: app/src/main/java/znyoo/name/baseproject/SampleApplicationLike.kt
package znyoo.name.baseproject

import android.app.Application
import android.content.Context
import android.content.Intent
import android.os.Environment
import android.util.Log
import android.view.View
import com.blankj.utilcode.util.LogUtils
import com.coder.zzq.smartshow.core.SmartShow
import com.scwang.smart.refresh.footer.BallPulseFooter
import com.scwang.smart.refresh.header.MaterialHeader
import com.scwang.smart.refresh.layout.SmartRefreshLayout
import com.tencent.bugly.Bugly
import com.tencent.bugly.beta.Beta
import com.tencent.bugly.beta.UpgradeInfo
import com.tencent.bugly.beta.download.DownloadListener
import com.tencent.bugly.beta.download.DownloadTask
import com.tencent.bugly.beta.ui.UILifecycleListener
import com.tencent.bugly.beta.upgrade.UpgradeListener
import com.tencent.bugly.beta.upgrade.UpgradeStateListener
import com.tencent.tinker.entry.DefaultApplicationLike
import me.jessyan.autosize.AutoSizeConfig
import znyoo.name.base.common.BUGLY_APPID
import znyoo.name.base.common.BaseConfig
import znyoo.name.base.common.ISTEST
import znyoo.name.base.common.initLog
import znyoo.name.netlibrary.retrofit.OnlineConfig
import znyoo.name.netlibrary.retrofit.XH_RXOnline
import znyoo.name.netlibrary.retrofit.api.A.MAIN_URL
import znyoo.name.netlibrary.retrofit.api.A.TEST_MAIN_URL


/**
 * @ProjectName:    BaseProject
 * @Package:        znyoo.name.baseproject
 * @ClassName:      SampliApplicationLike
 * @Author:         dhkyhb
 * @CreateDate:     2020/6/24 11:22 AM
 * @Description:     java类作用描述
 */
open class SampleApplicationLike(
    application: Application?, tinkerFlags: Int,
    tinkerLoadVerifyFlag: Boolean, applicationStartElapsedTime: Long,
    applicationStartMillisTime: Long, tinkerResultIntent: Intent?
) : DefaultApplicationLike(
    application, tinkerFlags,
    tinkerLoadVerifyFlag, applicationStartElapsedTime,
    applicationStartMillisTime, tinkerResultIntent
) {

    init {
        instance = this
    }

    //static 代码段可以防止内存泄露
    companion object {
        init {
            //设置全局的Header构建器
            SmartRefreshLayout.setDefaultRefreshHeaderCreator { context, layout -> MaterialHeader(context) }
            //设置全局的Footer构建器
            SmartRefreshLayout.setDefaultRefreshFooterCreator { context, layout -> BallPulseFooter(context) }
        }

        val TAG = "Tinker.SampleApplicationLike"
        private lateinit var instance: SampleApplicationLike
        fun getInstance(): SampleApplicationLike {
            return instance
        }
    }

    override fun onCreate() {
        super.onCreate()
        initialize()
//        inject(application as SampleApplication)
    }

    private fun initialize(){

        /**** Beta高级设置*****/
        /**
         * true表示app启动自动初始化升级模块；
         * false不好自动初始化
         * 开发者如果担心sdk初始化影响app启动速度，可以设置为false
         * 在后面某个时刻手动调用
         */
        Beta.autoInit = true

        /**
         * true表示初始化时自动检查升级
         * false表示不会自动检查升级，需要手动调用Beta.checkUpgrade()方法
         */
        Beta.autoCheckUpgrade = true

        /**
         * 设置升级周期为60s（默认检查周期为0s），60s内SDK不重复向后天请求策略
         */
        Beta.initDelay = 1 * 1000.toLong()

        /**
         * 设置通知栏大图标，largeIconId为项目中的图片资源；
         */
        Beta.largeIconId = R.mipmap.ic_launcher

        /**
         * 设置状态栏小图标，smallIconId为项目中的图片资源id;
         */
        Beta.smallIconId = R.mipmap.ic_launcher


        /**
         * 设置更新弹窗默认展示的banner，defaultBannerId为项目中的图片资源Id;
         * 当后台配置的banner拉取失败时显示此banner，默认不设置则展示“loading“;
         */
        Beta.defaultBannerId = R.mipmap.ic_launcher

        /**
         * 设置sd卡的Download为更新资源保存目录;
         * 后续更新资源会保存在此目录，需要在manifest中添加WRITE_EXTERNAL_STORAGE权限;
         */
        Beta.storageDir =
            Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)

        /**
         * 点击过确认的弹窗在APP下次启动自动检查更新时会再次显示;
         */
        Beta.showInterruptedStrategy = false

        /**
         * 只允许在MainActivity上显示更新弹窗，其他activity上不显示弹窗;
         * 不设置会默认所有activity都可以显示弹窗;
         */
        Beta.canShowUpgradeActs.add(MainActivity::class.java)


        /**
         *  设置自定义升级对话框UI布局
         *  注意：因为要保持接口统一，需要用户在指定控件按照以下方式设置tag，否则会影响您的正常使用：
         *  标题：beta_title，如：android:tag="beta_title"
         *  升级信息：beta_upgrade_info  如： android:tag="beta_upgrade_info"
         *  更新属性：beta_upgrade_feature 如： android:tag="beta_upgrade_feature"
         *  取消按钮：beta_cancel_button 如：android:tag="beta_cancel_button"
         *  确定按钮：beta_confirm_button 如：android:tag="beta_confirm_button"
         *  详见layout/upgrade_dialog.xml
         */
//        Beta.upgradeDialogLayoutId = R.layout.upgrade_dialog;

        /**
         * 设置自定义tip弹窗UI布局
         * 注意：因为要保持接口统一，需要用户在指定控件按照以下方式设置tag，否则会影响您的正常使用：
         *  标题：beta_title，如：android:tag="beta_title"
         *  提示信息：beta_tip_message 如： android:tag="beta_tip_message"
         *  取消按钮：beta_cancel_button 如：android:tag="beta_cancel_button"
         *  确定按钮：beta_confirm_button 如：android:tag="beta_confirm_button"
         *  详见layout/tips_dialog.xml
         */
//        Beta.tipsDialogLayoutId = R.layout.tips_dialog;

        /**
         *  如果想监听升级对话框的生命周期事件，可以通过设置OnUILifecycleListener接口
         *  回调参数解释：
         *  context - 当前弹窗上下文对象
         *  view - 升级对话框的根布局视图，可通过这个对象查找指定view控件
         *  upgradeInfo - 升级信息
         */
//        Beta.upgradeDialogLifecycleListener = object: UILifecycleListener<UpgradeInfo>() {
//
//            override fun onCreate(p0: Context?, p1: View?, p2: UpgradeInfo?) {
//                LogUtils.d(TAG, "onCreate")
//            }
//
//            override fun onResume(p0: Context?, p1: View?, p2: UpgradeInfo?) {
//                LogUtils.d(TAG, "onResume")
//            }
//
//            override fun onPause(p0: Context?, p1: View?, p2: UpgradeInfo?) {
//                LogUtils.d(TAG, "onPause")
//            }
//
//            override fun onStart(p0: Context?, p1: View?, p2: UpgradeInfo?) {
//                LogUtils.d(TAG, "onStart")
//            }
//
//            override fun onStop(p0: Context?, p1: View?, p2: UpgradeInfo?) {
//                LogUtils.d(TAG, "onStop")
//            }
//
//            override fun onDestroy(p0: Context?, p1: View?, p2: UpgradeInfo?) {
//                LogUtils.d(TAG, "onDestroy")
//            }
//        };

        /**
         * 自定义Activity参考，通过回调接口来跳转到你自定义的Actiivty中。
         */
//         Beta.upgradeListener = object: UpgradeListener {
//             override fun onUpgrade(p0: Int, p1: UpgradeInfo?, p2: Boolean, p3: Boolean) {
//                 if (p1 != null) {
//                     val i = Intent()
//                     i.setClass(application, MainActivity::class.java)
//                     i.flags = Intent.FLAG_ACTIVITY_NEW_TASK;
//                     application.startActivity(i)
//                 } else {
//
//                 }
//             }
//         }


        //监听安装包下载状态
        Beta.downloadListener =
            object : DownloadListener {
                override fun onReceive(downloadTask: DownloadTask) {
                    LogUtils.d(
                        "downloadListener receive apk file"
                    )
                }

                override fun onCompleted(downloadTask: DownloadTask) {
                    LogUtils.d(
                        "downloadListener download apk file success"
                    )
                }

                override fun onFailed(
                    downloadTask: DownloadTask,
                    i: Int,
                    s: String
                ) {
                    LogUtils.d(
                        "downloadListener download apk file fail"
                    )
                }
            }

        //监听APP升级状态

        //监听APP升级状态
        Beta.upgradeStateListener = object : UpgradeStateListener {
            override fun onUpgradeFailed(b: Boolean) {
                LogUtils.d(
                    "upgradeStateListener upgrade fail"
                )
            }

            override fun onUpgradeSuccess(b: Boolean) {
                LogUtils.d(
                    "upgradeStateListener upgrade success"
                )
            }

            override fun onUpgradeNoVersion(b: Boolean) {
                LogUtils.d(
                    "upgradeStateListener upgrade has no new version"
                )
            }

            override fun onUpgrading(b: Boolean) {
                LogUtils.d(
                    "upgradeStateListener upgrading"
                )
            }

            override fun onDownloadCompleted(b: Boolean) {
                LogUtils.d(
                    "upgradeStateListener download apk file success"
                )
            }
        }
        /**
         * 已经接入Bugly用户改用上面的初始化方法,不影响原有的crash上报功能;
         * init方法会自动检测更新，不需要再手动调用Beta.checkUpdate(),如需增加自动检查时机可以使用Beta.checkUpdate(false,false);
         * 参数1： applicationContext
         * 参数2：appId
         * 参数3：是否开启debug
         */
        Bugly.init(application, BUGLY_APPID, false)

        //rutosize 开启适配fragment
        AutoSizeConfig.getInstance().isCustomFragment = true;
        SmartShow.init(application)

        BaseConfig().init(application)
        initLog()
    }

}